### ISSUE TITLE
Need help on Distributed training using slurm ...

### ISSUE THREAD

**Issue by aadityachapagain** [facebookresearch/ParlAI#3084]
My Slurm setup: 
max_nodes: 8
GPU_count on each node: 8
GPU : v100

Training Script:
```bash 
#!/bin/sh
#SBATCH --job-name=distributed_example
#SBATCH --output=stdout.%j
#SBATCH --error=stderr.%j
#SBATCH --partition=gpu
#SBATCH --nodes=2
#SBATCH --time=0:12:00
#SBATCH --signal=SIGINT
#SBATCH --gres=gpu:8
#SBATCH --ntasks-per-node=8
#SBATCH --mem=80G
#SBATCH --cpus-per-task=2

srun python -u -m parlai.scripts.distributed_train \
-t blended_skill_talk,wizard_of_wikipedia,convai2:normalized \
--distributed-world-size 2 \
...
```
When i set --distributed-world-size to 2 . I got this error 
RuntimeError: NCCL error in: /pytorch/torch/lib/c10d/../c10d/NCCLUtils.hpp:82, invalid argument, NCCL version 2.4.8
srun: error: slurm-job-compute-0-0: tasks 2-7: Exited with exit code 1
srun: error: slurm-job-compute-0-1: tasks 8-13,15: Exited with exit code 1
srun: error: slurm-job-compute-0-1: task 14: Exited with exit code 1

But somehow  Training is still running and produce output. ( Which means primary tasks is still running , while all other secondary tasks are dead )

When i set --distributed-world-size to 16. I got this exception
RUNTIME ERROR: SLURM does not appear to be installed.

srun: error: slurm-job-compute-0-0: tasks 1-7: Exited with exit code 1
srun: error: slurm-job-compute-0-1: tasks 8-13: Exited with exit code 1

and Still Training is running and produce output . ( Which means Both primary tasks of both node 0 and 1 is running while all other secondary tasks are dead ).

**Comment by stephenroller** [comment]
The first NCCL error looks like a mismatch between the version of CUDA you have installed, and the version pytorch is trying to use.

The second comes from here:
https://github.com/facebookresearch/ParlAI/blob/cd20fd349873efd24b2266ce69a4f332ca0f88ae/parlai/utils/distributed.py#L388-L390

We spawn a subprocess to get a list of the nodes to set up distributed, and we call `scontrol` for that. For some reason, it can't find that command. You could maybe modify "scontrol" to the full unambiguous path.

Here's a slurm batch I launched I launched just yesterday

```bash
  #SBATCH --job-name=REDACTED_C
  #SBATCH --output=/REDACTED_A/slurm_logs/slrm_stdout.%j
  #SBATCH --error=/REDACTED_A/slurm_logs/slrm_stderr.%j
  #SBATCH --partition=REDACTED_B
  ## make sure we don't clobber log files if jobs get restarted
  #SBATCH --open-mode=append
  #SBATCH --nodes=1
  #SBATCH --time=24:00:00
  ## make sure we are told about preempts, and jobs running out of time, 60s beforehand
  #SBATCH --signal=USR1@60
  #SBATCH --cpus-per-task=5
  ## srun forks ntasks_per_node times on each node
  #SBATCH --ntasks-per-node=8
  #SBATCH --mem=200G
  #SBATCH --gpus-per-node=8

  if [[ "$SLURM_ARRAY_TASK_ID" == "0" ]]; then
    srun -K1 bash /REDACTED_A/grid001/run.sh >> /REDACTED_A/grid001/stdout.${SLURM_JOB_ID}_${SLURM_ARRAY_TASK_ID}
  fi
```

and inside that run.sh is (among general cluster specific setup):

```bash
  export NCCL_SOCKET_IFNAME=^docker0,lo
  export TOKENIZERS_PARALLELISM=true
  python -u -m parlai.scripts.distributed_train -mf /REDACTED_A/grid001/model OTHER_ARGS
```

**Comment by aadityachapagain** [comment]
Thanks for the quick reply. Just upgrading cuda10.1 to  cuda 10.2  fixed the issue.

**Comment by aadityachapagain** [comment]
Reopening this issues once again:

I am using GCP and created slurm cluster using https://github.com/SchedMD/slurm-gcp ,
Sbatch script
```
#!/bin/sh
#SBATCH --job-name=distributed_example
#SBATCH --output=stdout.%j
#SBATCH --error=stderr.%j
#SBATCH --partition=gpu
#SBATCH --nodes=2

#SBATCH --signal=SIGINT
#SBATCH --gpus-per-node=4
#SBATCH --ntasks-per-node=1
#SBATCH --mem=100G
#SBATCH --cpus-per-task=10

srun python -u -m parlai.scripts.distributed_train \
-t blended_skill_talk,wizard_of_wikipedia,convai2:normalized \
--distributed-world-size 2 \
--model-parallel true \
...
When I set #SBATCH --nodes=1 and --distributed-world-size 1. it works . It train model in one host using all available GPUs.
But when I set #SBATCH --nodes=2 and --distributed-world-size 2. Training get stuck at loading module.

here are some NCCL debug ,

slurm-train-compute-0-0:2139:2139 [0] NCCL INFO Bootstrap : Using [0]eth0:10.0.0.34<0>
slurm-train-compute-0-0:2139:2139 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
slurm-train-compute-0-0:2139:2139 [0] NCCL INFO NET/IB : No device found.
slurm-train-compute-0-0:2139:2139 [0] NCCL INFO NET/Socket : Using [0]eth0:10.0.0.34<0>
slurm-train-compute-0-0:2139:2139 [0] NCCL INFO Using network Socket
NCCL version 2.7.8+cuda10.2
slurm-train-compute-0-1:2119:2119 [1] NCCL INFO Bootstrap : Using [0]eth0:10.0.0.33<0>
slurm-train-compute-0-1:2119:2119 [1] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
slurm-train-compute-0-1:2119:2119 [1] NCCL INFO NET/IB : No device found.
slurm-train-compute-0-1:2119:2119 [1] NCCL INFO NET/Socket : Using [0]eth0:10.0.0.33<0>
slurm-train-compute-0-1:2119:2119 [1] NCCL INFO Using network Socket
slurm-train-compute-0-0:2139:2250 [0] NCCL INFO Channel 00/02 :    0   1
slurm-train-compute-0-0:2139:2250 [0] NCCL INFO Channel 01/02 :    0   1
slurm-train-compute-0-0:2139:2250 [0] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/64
slurm-train-compute-0-0:2139:2250 [0] NCCL INFO Trees [0] 1/-1/-1->0->-1|-1->0->1/-1/-1 [1] -1/-1/-1->0->1|1->0->-1/-1/-1
slurm-train-compute-0-1:2119:2212 [1] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/64
slurm-train-compute-0-1:2119:2212 [1] NCCL INFO Trees [0] -1/-1/-1->1->0|0->1->-1/-1/-1 [1] 0/-1/-1->1->-1|-1->1->0/-1/-1
slurm-train-compute-0-0:2139:2250 [0] NCCL INFO Channel 00 : 1[50] -> 0[40] [receive] via NET/Socket/0
slurm-train-compute-0-1:2119:2212 [1] NCCL INFO Channel 00 : 0[40] -> 1[50] [receive] via NET/Socket/0
slurm-train-compute-0-0:2139:2250 [0] NCCL INFO Channel 00 : 0[40] -> 1[50] [send] via NET/Socket/0
slurm-train-compute-0-1:2119:2212 [1] NCCL INFO Channel 00 : 1[50] -> 0[40] [send] via NET/Socket/0
slurm-train-compute-0-0:2139:2250 [0] NCCL INFO Channel 01 : 1[50] -> 0[40] [receive] via NET/Socket/0
slurm-train-compute-0-1:2119:2212 [1] NCCL INFO Channel 01 : 0[40] -> 1[50] [receive] via NET/Socket/0
slurm-train-compute-0-1:2119:2212 [1] NCCL INFO Channel 01 : 1[50] -> 0[40] [send] via NET/Socket/0
slurm-train-compute-0-0:2139:2250 [0] NCCL INFO Channel 01 : 0[40] -> 1[50] [send] via NET/Socket/0
slurm-train-compute-0-0:2139:2250 [0] NCCL INFO 2 coll channels, 2 p2p channels, 1 p2p channels per peer
slurm-train-compute-0-1:2119:2212 [1] NCCL INFO 2 coll channels, 2 p2p channels, 1 p2p channels per peer
slurm-train-compute-0-1:2119:2212 [1] NCCL INFO comm 0x7f2e38000e00 rank 1 nranks 2 cudaDev 1 busId 50 - Init COMPLETE
slurm-train-compute-0-0:2139:2250 [0] NCCL INFO comm 0x7f8130000e00 rank 0 nranks 2 cudaDev 0 busId 40 - Init COMPLETE
slurm-train-compute-0-0:2139:2139 [0] NCCL INFO Launch mode Parallel
<some random model info>
slurm-train-compute-0-0:2139:2386 [1] NCCL INFO Channel 00/02 :    0   1
slurm-train-compute-0-0:2139:2386 [1] NCCL INFO Channel 01/02 :    0   1
slurm-train-compute-0-0:2139:2386 [1] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/64
slurm-train-compute-0-0:2139:2386 [1] NCCL INFO Trees [0] 1/-1/-1->0->-1|-1->0->1/-1/-1 [1] -1/-1/-1->0->1|1->0->-1/-1/-1
slurm-train-compute-0-1:2119:2291 [0] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/64
slurm-train-compute-0-1:2119:2291 [0] NCCL INFO Trees [0] -1/-1/-1->1->0|0->1->-1/-1/-1 [1] 0/-1/-1->1->-1|-1->1->0/-1/-1
slurm-train-compute-0-1:2119:2291 [0] NCCL INFO Channel 00 : 0[50] -> 1[40] [receive] via NET/Socket/0
slurm-train-compute-0-0:2139:2386 [1] NCCL INFO Channel 00 : 1[40] -> 0[50] [receive] via NET/Socket/0
slurm-train-compute-0-1:2119:2291 [0] NCCL INFO Channel 00 : 1[40] -> 0[50] [send] via NET/Socket/0
slurm-train-compute-0-0:2139:2386 [1] NCCL INFO Channel 00 : 0[50] -> 1[40] [send] via NET/Socket/0
slurm-train-compute-0-1:2119:2291 [0] NCCL INFO Channel 01 : 0[50] -> 1[40] [receive] via NET/Socket/0
slurm-train-compute-0-0:2139:2386 [1] NCCL INFO Channel 01 : 1[40] -> 0[50] [receive] via NET/Socket/0
slurm-train-compute-0-1:2119:2291 [0] NCCL INFO Channel 01 : 1[40] -> 0[50] [send] via NET/Socket/0
slurm-train-compute-0-0:2139:2386 [1] NCCL INFO Channel 01 : 0[50] -> 1[40] [send] via NET/Socket/0
slurm-train-compute-0-0:2139:2386 [1] NCCL INFO 2 coll channels, 2 p2p channels, 1 p2p channels per peer
slurm-train-compute-0-1:2119:2291 [0] NCCL INFO 2 coll channels, 2 p2p channels, 1 p2p channels per peer
slurm-train-compute-0-0:2139:2386 [1] NCCL INFO comm 0x7f8060000e00 rank 0 nranks 2 cudaDev 1 busId 50 - Init COMPLETE
slurm-train-compute-0-1:2119:2291 [0] NCCL INFO comm 0x7f2de8000e00 rank 1 nranks 2 cudaDev 0 busId 40 - Init COMPLETE
slurm-train-compute-0-0:2139:2139 [1] NCCL INFO Launch mode Parallel

slurm-train-compute-0-1:2119:2292 [0] transport/net_socket.cc:414 NCCL WARN NET/Socket : message truncated : receiving 1048576 bytes instead of 524288
slurm-train-compute-0-1:2119:2292 [0] NCCL INFO include/net.h:28 -> 3
slurm-train-compute-0-1:2119:2292 [0] NCCL INFO transport/net.cc:357 -> 3
slurm-train-compute-0-1:2119:2292 [0] NCCL INFO proxy.cc:198 -> 3 [Proxy Thread]

Logs in node-0:
[2020-12-30T05:11:27.540] error: Domain socket directory /var/spool/slurmd: No such file or directory
[2020-12-30T05:11:27.552] error: xcpuinfo_hwloc_topo_load: failed (load will be required after read failures).
[2020-12-30T05:11:27.552] error: Node configuration differs from hardware: Procs=16:16(hw) Boards=1:1(hw) SocketsPerBoard=16:1(hw) CoresPerSocket=1:8(hw) ThreadsPerCore=1:2(hw)
[2020-12-30T05:11:27.552] Message aggregation disabled
[2020-12-30T05:11:27.605] CPU frequency setting not configured for this node
[2020-12-30T05:11:27.627] slurmd version 19.05.8 started
[2020-12-30T05:11:27.642] slurmd started on Wed, 30 Dec 2020 05:11:27 +0000
[2020-12-30T05:11:27.642] CPUs=16 Boards=1 Sockets=16 Cores=1 Threads=1 Memory=60231 TmpDisk=409388 Uptime=22 CPUSpecList=(null) FeaturesAvail=(null) FeaturesActive=(null)
[2020-12-30T05:11:51.871] task_p_slurmd_batch_request: 25
[2020-12-30T05:11:51.871] task/affinity: job 25 CPU input mask for node: 0x03FF
[2020-12-30T05:11:51.871] task/affinity: job 25 CPU final HW mask for node: 0x1F1F
[2020-12-30T05:11:51.872] _run_prolog: run job script took usec=216
[2020-12-30T05:11:51.872] _run_prolog: prolog with lock for job 25 ran for 0 seconds
[2020-12-30T05:11:51.873] Launching batch job 25 for UID 3521375391
[2020-12-30T05:11:51.937] [25.batch] task/cgroup: /slurm/uid_3521375391/job_25: alloc=57344MB mem.limit=57344MB memsw.limit=57344MB
[2020-12-30T05:11:51.943] [25.batch] task/cgroup: /slurm/uid_3521375391/job_25/step_batch: alloc=57344MB mem.limit=57344MB memsw.limit=57344MB
[2020-12-30T05:11:51.960] [25.batch] task_p_pre_launch: Using sched_affinity for tasks
[2020-12-30T05:11:52.045] launch task 25.0 request from UID:3521375391 GID:3521375391 HOST:10.0.0.34 PORT:53419
[2020-12-30T05:11:52.045] lllp_distribution jobid [25] implicit auto binding: sockets,one_thread, dist 1
[2020-12-30T05:11:52.045] _task_layout_lllp_cyclic 
[2020-12-30T05:11:52.045] _lllp_generate_cpu_bind jobid [25]: mask_cpu,one_thread, 0x1F1F
[2020-12-30T05:11:52.057] [25.0] task/cgroup: /slurm/uid_3521375391/job_25: alloc=57344MB mem.limit=57344MB memsw.limit=57344MB
[2020-12-30T05:11:52.062] [25.0] task/cgroup: /slurm/uid_3521375391/job_25/step_0: alloc=57344MB mem.limit=57344MB memsw.limit=57344MB
[2020-12-30T05:11:52.073] [25.0] task_p_pre_launch: Using sched_affinity for tasks

Logs in node-1:
[2020-12-30T05:11:27.384] error: Domain socket directory /var/spool/slurmd: No such file or directory
[2020-12-30T05:11:27.400] error: xcpuinfo_hwloc_topo_load: failed (load will be required after read failures).
[2020-12-30T05:11:27.400] error: Node configuration differs from hardware: Procs=16:16(hw) Boards=1:1(hw) SocketsPerBoard=16:1(hw) CoresPerSocket=1:8(hw) ThreadsPerCore=1:2(hw)
[2020-12-30T05:11:27.400] Message aggregation disabled
[2020-12-30T05:11:27.437] CPU frequency setting not configured for this node
[2020-12-30T05:11:27.456] slurmd version 19.05.8 started
[2020-12-30T05:11:27.468] slurmd started on Wed, 30 Dec 2020 05:11:27 +0000
[2020-12-30T05:11:27.468] CPUs=16 Boards=1 Sockets=16 Cores=1 Threads=1 Memory=60231 TmpDisk=409388 Uptime=20 CPUSpecList=(null) FeaturesAvail=(null) FeaturesActive=(null)
[2020-12-30T05:11:52.048] launch task 25.0 request from UID:3521375391 GID:3521375391 HOST:10.0.0.34 PORT:50381
[2020-12-30T05:11:52.048] lllp_distribution jobid [25] implicit auto binding: sockets,one_thread, dist 1
[2020-12-30T05:11:52.048] _task_layout_lllp_cyclic 
[2020-12-30T05:11:52.048] _lllp_generate_cpu_bind jobid [25]: mask_cpu,one_thread, 0x1F1F
[2020-12-30T05:11:52.051] _run_prolog: run job script took usec=2614
[2020-12-30T05:11:52.051] _run_prolog: prolog with lock for job 25 ran for 0 seconds
[2020-12-30T05:11:52.097] [25.0] task/cgroup: /slurm/uid_3521375391/job_25: alloc=57344MB mem.limit=57344MB memsw.limit=57344MB
[2020-12-30T05:11:52.103] [25.0] task/cgroup: /slurm/uid_3521375391/job_25/step_0: alloc=57344MB mem.limit=57344MB memsw.limit=57344MB
[2020-12-30T05:11:52.115] [25.0] task_p_pre_launch: Using sched_affinity for tasks

Any solution to this would really be helpful .
Thank you.

**Comment by stephenroller** [comment]
Do the stdout logs say something about initializing on rank X? Can you add --verbose to the parlai call and paste those logs?

Eyeballing the config looks right to me.

My only thought is maybe the nodes are having trouble connecting to each other. Maybe they're trying to connect to the wrong hosts? Not sure

**Comment by aadityachapagain** [comment]
2020-12-30 05:11:59,533 INFO     | Initializing host slurm-train-compute-0-1 as rank 1, main is slurm-train-compute-0-0
2020-12-30 05:12:00,947 INFO     | Initializing host slurm-train-compute-0-0 as rank 0, main is slurm-train-compute-0-0
rank:  0 | 2020-12-30 06:29:10,696 INFO     | Distributed group initialized
rank:  1 | 2020-12-30 06:29:10,696 INFO     | Distributed group initialized
rank:  0 | 2020-12-30 06:29:10,954 INFO     | building dictionary first...
rank:  0 | 2020-12-30 06:29:10,959 WARNING  | <REDACTED>
rank:  0 | 2020-12-30 06:29:10,960 WARNING  | <REDACTED>
rank:  0 | 2020-12-30 06:29:10,960 WARNING  | <REDACTED>
rank:  0 | 2020-12-30 06:29:10,960 WARNING  | <REDACTED>
rank:  0 | 2020-12-30 06:29:10,961 WARNING  | <REDACTED>
rank:  0 | 2020-12-30 06:29:10,961 WARNING  | <REDACTED>
rank:  0 | 2020-12-30 06:29:10,961 WARNING  | <REDACTED>
rank:  0 | 2020-12-30 06:29:10,961 WARNING  | <REDACTED>
rank:  0 | 2020-12-30 06:29:10,961 WARNING  | <REDACTED>
initializing the weights with: distributed_world_size: 2,verbose: True,download_path: Non
e,datapath: <REDACTED>/ParlAI/data,load_from_checkpoint: True,port: 61337,interactive_mode: False,rank: 0
rank:  0 | 2020-12-30 06:29:10,971 INFO     | Using CUDA
rank:  0 | 2020-12-30 06:29:10,974 INFO     | loading dictionary from <REDACTED>

Training stuck here forever ...

**Comment by stephenroller** [comment]
Oh have you checked stderr? I believe dictionaries must be prebuilt to use with distributed. You can use `parlai build_dict` to make one, or use the gpt2 tokenizer.

**Comment by aadityachapagain** [comment]
stderr is empty.
thinking dictionaries are not problem here, as it able to build dictionary in node.

**Comment by aadityachapagain** [comment]
difference between When running sbatch in more nodes and 1 nodes, I found is When running more nodes It run NCCL with mode Parallel and rank 0 node hung up,

NCCL debug:
slurm-train-compute-0-0:2115:2265 [1] NCCL INFO Channel 00/02 :    0   1
slurm-train-compute-0-0:2115:2265 [1] NCCL INFO Channel 01/02 :    0   1
slurm-train-compute-0-1:2104:2277 [0] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/64
slurm-train-compute-0-1:2104:2277 [0] NCCL INFO Trees [0] -1/-1/-1->1->0|0->1->-1/-1/-1 [1] 0/-1/-1->1->-1|-1->1->0/-1/-1
slurm-train-compute-0-0:2115:2265 [1] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/64
slurm-train-compute-0-0:2115:2265 [1] NCCL INFO Trees [0] 1/-1/-1->0->-1|-1->0->1/-1/-1 [1] -1/-1/-1->0->1|1->0->-1/-1/-1
slurm-train-compute-0-1:2104:2277 [0] NCCL INFO Channel 00 : 0[50] -> 1[40] [receive] via NET/Socket/0
slurm-train-compute-0-0:2115:2265 [1] NCCL INFO Channel 00 : 1[40] -> 0[50] [receive] via NET/Socket/0
slurm-train-compute-0-1:2104:2277 [0] NCCL INFO Channel 00 : 1[40] -> 0[50] [send] via NET/Socket/0
slurm-train-compute-0-0:2115:2265 [1] NCCL INFO Channel 00 : 0[50] -> 1[40] [send] via NET/Socket/0
slurm-train-compute-0-1:2104:2277 [0] NCCL INFO Channel 01 : 0[50] -> 1[40] [receive] via NET/Socket/0
slurm-train-compute-0-0:2115:2265 [1] NCCL INFO Channel 01 : 1[40] -> 0[50] [receive] via NET/Socket/0
slurm-train-compute-0-1:2104:2277 [0] NCCL INFO Channel 01 : 1[40] -> 0[50] [send] via NET/Socket/0
slurm-train-compute-0-0:2115:2265 [1] NCCL INFO Channel 01 : 0[50] -> 1[40] [send] via NET/Socket/0
slurm-train-compute-0-1:2104:2277 [0] NCCL INFO 2 coll channels, 2 p2p channels, 1 p2p channels per peer
slurm-train-compute-0-0:2115:2265 [1] NCCL INFO 2 coll channels, 2 p2p channels, 1 p2p channels per peer
slurm-train-compute-0-0:2115:2265 [1] NCCL INFO comm 0x7f0a58000e00 rank 0 nranks 2 cudaDev 1 busId 50 - Init COMPLETE
slurm-train-compute-0-0:2115:2115 [1] NCCL INFO Launch mode Parallel
slurm-train-compute-0-1:2104:2277 [0] NCCL INFO comm 0x7fd968000e00 rank 1 nranks 2 cudaDev 0 busId 40 - Init COMPLETE

slurm-train-compute-0-1:2104:2278 [0] transport/net_socket.cc:414 NCCL WARN NET/Socket : message truncated : receiving 1048576 bytes instead of 524288
slurm-train-compute-0-1:2104:2278 [0] NCCL INFO include/net.h:28 -> 3
slurm-train-compute-0-1:2104:2278 [0] NCCL INFO transport/net.cc:357 -> 3
slurm-train-compute-0-1:2104:2278 [0] NCCL INFO proxy.cc:198 -> 3 [Proxy Thread]

You can see at first there is p2p communication between rank 0 node and rank 1 node,  but after rank 0 node launch NCCL in model Parallel,
rank 1 receive bytes and then system hungs , no more communication ...

While in the case with nodes=1,
There is no NCCL debug info that tells NCCL INFO Launch mode Parallel
...

**Comment by stephenroller** [comment]
Is it possible that the files being loaded are being loaded aren't accessible by both nodes? Can you try without --init-model?

Can you try with a small model to eliminate the possibility of memory issues?

**Comment by stephenroller** [comment]
They're getting past the initialization step, so they are finding each other.

The dictionary output still looks a tad off to me but I'm AFK to not know what's right.

**Comment by aadityachapagain** [comment]
Able to get Error after setting NCCL_ASYNC_ERROR_HANDLING=1
stderr output:
terminate called after throwing an instance of 'std::runtime_error'
  what():  NCCL error: internal error, NCCL version 2.7.8
Fatal Python error: Aborted

Thread 0x00007fb1ec38e740 (most recent call first):
  File "REDACTED/ParlAI/parlai/utils/distributed.py", line 221 in sync_object
  File "REDACTED/ParlAI/parlai/utils/distributed.py", line 335 in distributed_context
  File "REDACTED/anaconda3/lib/python3.8/contextlib.py", line 113 in __enter__
  File "REDACTED/ParlAI/parlai/utils/distributed.py", line 392 in slurm_distributed_context
  File "REDACTED/anaconda3/lib/python3.8/contextlib.py", line 113 in __enter__
  File "parlai/scripts/distributed_train.py", line 53 in run
  File "REDACTED/ParlAI/parlai/core/script.py", line 89 in _run_from_parser_and_opt
  File "REDACTED/ParlAI/parlai/core/script.py", line 83 in _run_args
  File "REDACTEDParlAI/parlai/core/script.py", line 110 in main
  File "parlai/scripts/distributed_train.py", line 59 in <module>
model-train.sh: line 40:  2093 Aborted                 python parlai/scripts/distributed_train.py -t REDACTED
terminate called after throwing an instance of 'std::runtime_error'
  what():  NCCL error: unhandled system error, NCCL version 2.7.8
Fatal Python error: Aborted

Thread 0x00007fd9e27af740 (most recent call first):
  File "REDACTED/anaconda3/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 954 in all_reduce
  File "REDACTED/ParlAI/parlai/utils/distributed.py", line 255 in sync_parameters
  File "REDACTED/ParlAI/parlai/core/torch_generator_agent.py", line 516 in __init__
  File "REDACTED/ParlAI/parlai/core/agents.py", line 419 in create_agent_from_opt_file
  File "REDACTED/ParlAI/parlai/core/agents.py", line 532 in create_agent
  File "REDACTED/ParlAI/parlai/scripts/train_model.py", line 317 in __init__
  File "parlai/scripts/distributed_train.py", line 54 in run
  File "REDACTED/ParlAI/parlai/core/script.py", line 89 in _run_from_parser_and_opt
  File "REDACTED/ParlAI/parlai/core/script.py", line 83 in _run_args
  File "REDACTED/ParlAI/parlai/core/script.py", line 110 in main
  File "parlai/scripts/distributed_train.py", line 59 in <module>
srun: error: slurm-train-compute-0-1: task 1: Exited with exit code 134
srun: Terminating job step 39.0
model-train.sh: line 40:  2112 Aborted                 python parlai/scripts/distributed_train.py -t REDACTED
srun: error: slurm-train-compute-0-0: task 0: Exited with exit code 134

seems like, NCCL hung up during syncing parameters ...

**Comment by aadityachapagain** [comment]
> Is it possible that the files being loaded are being loaded aren't accessible by both nodes? Can you try without --init-model?
> 
> Can you try with a small model to eliminate the possibility of memory issues?

All new spawning compute nodes during batch  job, use shared file system, so every files are accessible, I still checked and it is accessible no exception on that,
and I tried 90M model still same issue, when using more than 1 nodes...

**Comment by aadityachapagain** [comment]
Thanks for your support @stephenroller , NCCL was the source of the error as I suspected, It is running on **NCCL_LAUNCH_MODE=PARALLEL**, tried to set **NCCL_LAUNCH_MODE=GROUP**, inside sbatch script, but seems like it is forced to run in PARALLEL model. 

Switched backend from *nccl* to *gloo*, Now, its working ,but training is very slow, 
Created new issue on [Nvidia/Nccl](https://github.com/NVIDIA/nccl/issues/447)

**Comment by stephenroller** [comment]
Double check all your versions etc.

I work mostly on a managed cluster so I've not particularly run into this before...

**Comment by mesllo** [comment]
@aadityachapagain did you ever figure out what the problem was for you?
